### PR TITLE
Arya game

### PR DESCRIPTION
**Pull request by aryanagarkar**

### CODE REVIEW

File: HasseGraph/service/Graph.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿using Service.Util;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace service
+{
+    public class Graph
+    {
+        readonly MDictionary<string, GraphNode> idNodeMap;
+        List<GraphLink> links = new MList<GraphLink>();
```
Comment (nagarkar) [inline HasseGraph/service/Graph.cs:10]: Why isn't this readonly?

File: HasseGraph/service/Graph.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+using System.Linq;
+
+namespace service
+{
+    public class Graph
+    {
+        readonly MDictionary<string, GraphNode> idNodeMap;
+        List<GraphLink> links = new MList<GraphLink>();
+
+        public Graph(MDictionary<string, GraphNode> idNodeMap, List<GraphLink> links)
```
Comment (nagarkar) [inline HasseGraph/service/Graph.cs:12]: Why is this public?

File: HasseGraph/service/Graph.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        public MDictionary<string, GraphNode> IdNodeMap
+        {
+            get { return idNodeMap; }
+        }
+
+        public List<GraphNode> Nodes{
+            get{return idNodeMap.Values.ToList(); }
+        }
+
+        public  List<GraphLink> Links{
```
Comment (nagarkar) [inline HasseGraph/service/Graph.cs:27]: Please format all files. I see needless spaces and lack of spaces etc.

File: HasseGraph/service/GraphBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+﻿using Service.Util;
+using System.Collections.Generic;
+
+namespace service
+{
+    public class GraphBuilder
+    {
+        MDictionary<string, GraphNode> idNodeMap = new MDictionary<string, GraphNode>();
```
Comment (nagarkar) [inline HasseGraph/service/GraphBuilder.cs:8]: Should these be private or is private the default?

File: HasseGraph/service/GraphBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    public class GraphBuilder
+    {
+        MDictionary<string, GraphNode> idNodeMap = new MDictionary<string, GraphNode>();
+        List<GraphLink> links = new MList<GraphLink>();
+
+        public GraphBuilder()
+        {  
+        }
+
+        GraphBuilder(List<GraphLink> links, MDictionary<string, GraphNode> idNodeMap){
```
Comment (nagarkar) [inline HasseGraph/service/GraphBuilder.cs:15]: Why do you need this?

File: HasseGraph/service/GraphBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            this.links = links;
+            this.idNodeMap = idNodeMap;
+        }
+
+        public GraphBuilder addLink(string id1, string id2){
+            GraphNode nodeId1 = addOrGetNode(id1);
+            GraphNode nodeId2 = addOrGetNode(id2);
+            links.Add(new GraphLink(nodeId1, nodeId2));
```
Comment (nagarkar) [inline HasseGraph/service/GraphBuilder.cs:23]: What if this link is already present? Where are you checking you don't add it twice?

File: HasseGraph/service/GraphBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+            
+            nodeId1.OutgoingLinks.Add(nodeId2);
+            nodeId2.IncomingLinks.Add(nodeId1);
+
+            return this;
+        }
+
+        public Graph build(){
```
Comment (nagarkar) [inline HasseGraph/service/GraphBuilder.cs:31]: Public methods at top, private at bottom.

File: HasseGraph/service/GraphLink.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
+﻿using System.Collections.Generic;
+
+namespace service
+{
+    public class GraphLink
+    {
+        GraphNode source;
```
Comment (nagarkar) [inline HasseGraph/service/GraphLink.cs:7]: ditto question

File: HasseGraph/service/GraphNode.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+{
+    public class GraphNode
+    {
+        readonly List<GraphNode> incomingLinks;
+        readonly List<GraphNode> outgoingLinks;
+
+
+        public GraphNode()
+        {
+            incomingLinks = new MList<GraphNode>();
```
Comment (nagarkar) [inline HasseGraph/service/GraphNode.cs:14]: I don't see an MList class defined anywhere.

**Review by nagarkar — CHANGES_REQUESTED** [review]

File: HasseGraph/service/Graph.cs (hunk 1 above)
Comment (aryanagarkar) [reply HasseGraph/service/Graph.cs:10]: Changed this to read-only.

File: HasseGraph/service/Graph.cs (hunk 3 above)
Comment (aryanagarkar) [reply HasseGraph/service/Graph.cs:27]: Formatted all the files.

File: HasseGraph/service/GraphBuilder.cs (hunk 4 above)
Comment (aryanagarkar) [reply HasseGraph/service/GraphBuilder.cs:8]: If you don't specify, these are private by default.

File: HasseGraph/service/GraphBuilder.cs (hunk 5 above)
Comment (aryanagarkar) [reply HasseGraph/service/GraphBuilder.cs:15]: I changed the other code and I didn't need this but I forgot to delete it. I deleted it.

File: HasseGraph/service/GraphBuilder.cs (hunk 6 above)
Comment (aryanagarkar) [reply HasseGraph/service/GraphBuilder.cs:23]: I changed the code to check if the link is already present.

File: HasseGraph/service/GraphNode.cs (hunk 9 above)
Comment (aryanagarkar) [reply HasseGraph/service/GraphNode.cs:14]: I forgot to commit that class, I will commit it in the next commit.

File: HasseGraph/service/GraphBuilder.cs (hunk 7 above)
Comment (aryanagarkar) [reply HasseGraph/service/GraphBuilder.cs:31]: I fixed all the files too have public methods on top, and private on the bottom.

File: HasseGraph/service/Graph.cs (hunk 2 above)
Comment (aryanagarkar) [reply HasseGraph/service/Graph.cs:12]: If I make all of these classes private, I will need to put them into nested classes and one file would become huge. I am going to try to make it internal instead.

File: HasseGraph/service/GraphLink.cs (hunk 8 above)
Comment (aryanagarkar) [reply HasseGraph/service/GraphLink.cs:7]: If I make all of these classes private, I will need to put them into nested classes and one file would become huge. I am going to try to make it internal instead.

**Pull request closed without merging**